File: 146-lru-cache/lru-cache.cpp
class Node{
 public:
    int key;
    int value;
    Node* prev = NULL;
    Node* next = NULL;

    Node(int key, int val){
        this->key = key;
        this->value = val;
    }
    int getValue(){
        return value;
    }

};
class LRUCache {
    Node* start;
    Node* end;
    unordered_map<int,Node*> m;
    int capacity;
public:
    LRUCache(int capacity) {
        start = new Node(-1,-1);
        end = new Node(-1,-1);
        start->next = end;
        end->prev = start;
        this->capacity = capacity;
    }
    
    int get(int key) {
        if(m.find(key)!=m.end()){
           Node* temp = m[key];
           deleteNode(key);  
           addNodeAtFront(temp);
           m[key] = temp;
           return temp->value;
        }
        else return -1;
    }
    void addNodeAtFront(Node* temp)
    {
        
        temp->next = start->next;
        start->next->prev = temp;
        start->next = temp;
        temp->prev = start;

    }
    void deleteNode(int key){
        Node* temp = m[key];
        m[key]->prev->next =temp->next;
        m[key]->next->prev = temp->prev;
    }
    void put(int key, int val) {
        if(m.find(key)!=m.end())
        {  
            Node* temp = m[key];
            temp->value = val;
            deleteNode(key);
            addNodeAtFront(temp);
            m[key] = temp;
        }
        else if(capacity > m.size()){
            Node* temp = new Node(key,val);
            addNodeAtFront(temp);
            m[key] = temp;
        }
        else{
            int k = end->prev->key;
            deleteNode(end->prev->key);
            m.erase(k);

            Node* temp = new Node(key,val);
            addNodeAtFront(temp);
            m[key] = temp;
        }
        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 146-lru-cache/lru-cache.java
class Node{
    int key;
    int val;
    Node next,prev;
    Node(int key,int val){
         this.key = key;
         this.val = val;
    }

}
class LRUCache {
    int cap;
    int size;
    Map<Integer,Node> mp;
    Node start, end;
    public LRUCache(int capacity) {
       cap = capacity;
       size = 0;
       mp = new HashMap<>();
       start = new Node(0,0);
       end = new Node(0,0);
       start.next = end;
       end.prev = start;
    }
    
    public int get(int key) {
        if(!mp.containsKey(key)) return -1;
        else{
            Node temp = mp.get(key);
            removeNode(temp);
            addNodeFront(temp);
            return temp.val;
        }
    }
    
    public void put(int key, int value) {
        if(!mp.containsKey(key))
        {
            Node temp = new Node(key,value);
            mp.put(key,temp);
            addNodeFront(temp);
            if(size<cap)
            {
                 size++;
            }
            else{
                Node delete = end.prev;
                mp.remove(delete.key);
                removeNode(delete);
            }
        }
        else{
            Node temp = mp.get(key);
            temp.val = value;
            removeNode(temp);
            addNodeFront(temp);
        }
    }
    
    public void removeNode(Node curr)
    {
        Node n = curr.next;
        Node p = curr.prev;
        p.next = n;
        n.prev = p;
    }
    public void addNodeFront(Node curr)
    {
        Node n = start.next;
        start.next = curr;
        curr.prev = start;
        curr.next = n;
        n.prev = curr;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */